### PR TITLE
Update some models for AMP training

### PR DESCRIPTION
**Pull request by epwalsh**
Fixes models that use an RNN cell by wrapping the forward call to the cell within an `autocast(False)` context.

### CODE REVIEW

**Comment by epwalsh** [comment]
Current failure may be related to https://github.com/pytorch/pytorch/issues/36428, but note that we're actually using `LSMTCell` directly here and still seeing a failure, despite [this comment](https://github.com/pytorch/pytorch/issues/42229#issuecomment-665740611) which says AMP should work fine when using just RNN cells.

File: allennlp_models/generation/models/copynet_seq2seq.py [hunk 1]
```diff
@@ -336,9 +336,13 @@ def _decoder_step(
         # shape: (group_size, decoder_input_dim)
         projected_decoder_input = self._input_projection_layer(decoder_input)
 
-        state["decoder_hidden"], state["decoder_context"] = self._decoder_cell(
-            projected_decoder_input, (state["decoder_hidden"], state["decoder_context"])
-        )
+        # TODO (epwalsh): remove the autocast(False) once torch's AMP is working for RNNs.
```
Comment (matt-gardner) [inline allennlp_models/generation/models/copynet_seq2seq.py:339]: Maybe make this more specific?  Same with the other files.
```suggestion
        # TODO (epwalsh): remove the autocast(False) once torch's AMP is working for LSTMCells.
```

File: tests/generation/models/copynet_test.py [hunk 2]
```diff
@@ -16,9 +18,35 @@ def setup_method(self):
...
 
+    @requires_gpu
+    def test_model_can_train_with_amp(self):
+        train_model_from_file(
+            self.param_file,
+            self.TEST_DIR,
+            overrides="{'trainer.use_amp':true,'trainer.cuda_device':0}",
+        )
+
+        # NOTE: as of writing this test, AMP does not work with RNNs. Hence we had
```
Comment (matt-gardner) [inline tests/generation/models/copynet_test.py:32]: Again here, holdover from earlier.

Comment (epwalsh) [inline tests/generation/models/copynet_test.py:32]:
```suggestion
        # NOTE: as of writing this test, AMP does not work with RNNs and LSTMCells. Hence we had
```

File: allennlp_models/generation/models/simple_seq2seq.py [hunk 3]
```diff
@@ -433,9 +433,12 @@ def _prepare_output_projections(
 
         # shape (decoder_hidden): (batch_size, decoder_output_dim)
         # shape (decoder_context): (batch_size, decoder_output_dim)
-        decoder_hidden, decoder_context = self._decoder_cell(
-            decoder_input, (decoder_hidden, decoder_context)
-        )
+
+        # TODO (epwalsh): remove the autocast(False) once torch's AMP is working for RNNs.
```
Comment (epwalsh) [inline allennlp_models/generation/models/simple_seq2seq.py:437]:
```suggestion
        # TODO (epwalsh): remove the autocast(False) once torch's AMP is working for LSTMCells.
```

File: allennlp_models/generation/modules/decoder_nets/lstm_cell.py [hunk 4]
```diff
@@ -126,9 +126,11 @@ def forward(
 
         # shape (decoder_hidden): (batch_size, decoder_output_dim)
         # shape (decoder_context): (batch_size, decoder_output_dim)
-        decoder_hidden, decoder_context = self._decoder_cell(
-            decoder_input, (decoder_hidden, decoder_context)
-        )
+        # TODO (epwalsh): remove the autocast(False) once torch's AMP is working for RNNs.
```
Comment (epwalsh) [inline allennlp_models/generation/modules/decoder_nets/lstm_cell.py:129]:
```suggestion
        # TODO (epwalsh): remove the autocast(False) once torch's AMP is working for LSMTCells.
```

**Pull request merged**